source/extensions/filters/http/cors/cors_filter.cc

```diff
@@ -125,6 +131,12 @@ Http::FilterHeadersStatus CorsFilter::decodeHeaders(Http::RequestHeaderMap& head
     response_headers->setInline(access_control_max_age_handle.handle(), maxAge());
   }
 
+  if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.cors_private_network_access")) {
+    if (headers.getInlineValue(access_control_request_private_network_handle.handle()) == "true") {
+      response_headers->setInline(access_control_response_private_network_handle.handle(), "true");
```

Comment:
According to https://developer.chrome.com/blog/private-network-access-preflight/#server-side-requests:
> When your server receives a preflight request (an OPTIONS request with CORS headers), the server should check for the presence of an Access-Control-Request-Private-Network: true header. If this header is present on the request, the server should examine the Origin header and the request path along with any other relevant information (such as Access-Control-Request-Headers) to ensure the request is safe to allow. Typically, you should allow access to a single origin under your control.

I think we do not need to check the destination is more private than the origin since:
1. if it has this header, which means it knows it is requesting one more private network, and we have checked the Origin header and the request path along with any other relevant information (such as Access-Control-Request-Headers), we can just ensure it is safe to allow and send the response header
2. if it has not, the response header added or not become meaningless


> Chrome will start sending a CORS preflight request ahead of any [private network request](https://developer.chrome.com/blog/private-network-access-preflight/#what-is-private-network-access) for a subresource, which asks for explicit permission from the target server. This preflight request will carry a new header, Access-Control-Request-Private-Network: true, and the response to it must carry a corresponding header, Access-Control-Allow-Private-Network: true.

Now obviously this CORS header is added by **Chrome**  only for verification. I think this pair of headers is an extra check. I am not the CORS expert, if any misunderstandings, feel free to point them out, thanks.

Reply:
Not an expert on CORS as well.
If I read correctly the [spec proposal](https://wicg.github.io/private-network-access/#private-network-request-heading), I think that by sending back the `Access-Control-Allow-Private-Network: true` header, it implies that the server allows the resource to be shared with external networks, and that might not be the right thing to do.

@mattklein123 can you recommend a reviewer that can take a look at CORS from the server perspective?

Reply:
> @mattklein123 can you recommend a reviewer that can take a look at CORS from the server perspective?

Not really. @RyanTheOptimist is probably our best bet as someone that is more familiar with web standards?

Reply:
It looks like the implementation has changed since this discussion started. Initially there was a runtime feature to control sending `Access-Control-Allow-Private-Network: true` in response to `Access-Control-Request-Private-Network: true`. I agree that this does not seem correct as it would not allow a server to prevent public access to resources that should otherwise be private access only. However, the implementation now appears to check a CorsPolicy allow_private_network_access value, and only if that is true will it send `Access-Control-Allow-Private-Network: true`. This behavior sounds correct to me.

Reply:
IIRC the previous implementation which I commented on was essentially checking if the `Access-Control-Request-Private-Network: true` is sent as part of the request, and then automatically added `Access-Control-Allow-Private-Network: true` in the response. The question was whether there should be any other validation, because it seems to me that just adding a response to a request doesn't really help with anything.

Reply:
Does the current implementation seem correct to you? I would expect that by default, allow_private_network_access would be false and hence `Access-Control-Allow-Private-Network: true` would not be sent, and hence Chrome would block requests to this origin. Does that accurately capture the current implementation? Are you thinking that perhaps we should have a config which specifies a list of origins for which private network access it allowed and check the Origin: header against that list?

Reply:
@RyanTheOptimist yep, it is what exactly implementation did.

@adisuissa wait for your input :)